### PR TITLE
OnPrem Users

### PR DESCRIPTION
**Pull request by Basit-qc**
- Supporting onPrem customers
- Introducing on_prem_config.json that will point to OnPrem resources (Routes and URLs etc)
- Picking ApiRoutes dynamically, separated files for dev/staging/prod configs

### CODE REVIEW

File: src/config/dev.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+export const devConfig = {
```
Comment (saadberry) [inline src/config/dev.ts:1]: :+1:

File: src/utils/setup_utils.ts [hunk 2]
```diff
@@ -307,3 +310,31 @@ export const uuidv4 = () => {
...
+		{
+			ROOT_REPO: "~/.codesync",
+			API_HOST: "https://api.codesync.com",
+			API_BASE_URL: "https://api.codesync.com/v1",
+			SOCKET_HOST: "wss://api.codesync.com",
+			WEBAPP_HOST: "https://www.codesync.com",
+			CW_LOGS_GROUP: "client-logs",
+			AWS_REGION: "us-east-1"
+		};
+	*/
```
Comment (saadberry) [inline src/utils/setup_utils.ts:325]: :+1:

File: src/utils/url_utils.ts [hunk 3]
```diff
@@ -42,8 +43,24 @@ export const generateServerUrl = (urlPath: string, baseUrl=API_BASE_URL, addTime
...
-	const url = `${WEB_APP_URL}${urlPath}`;
+export const generateApiUrl = (urlPath: string, addTimezone=false) => {
+	const apiBaseUrl = getSystemConfig().API_BASE_URL;
+	return generateBackendUrl(urlPath, apiBaseUrl, addTimezone);
+};
+
+export const generateApiHostUrl = (urlPath: string) => {
+	const baseUrl = getSystemConfig().API_HOST;
+	return generateBackendUrl(urlPath, baseUrl);
+};
```
Comment (saadberry) [inline src/utils/url_utils.ts:54]: Host URL will be used for on prem customers?

Comment (Basit-qc) [reply src/utils/url_utils.ts:54]: Will be used generally, its value will be different for onPrem users and will point to their resources

File: src/extension.ts [hunk 4]
```diff
@@ -97,14 +98,9 @@ export async function activate(context: vscode.ExtensionContext) {
...
-					if (!changeEvent.changed[0]?.input ||
-						// @ts-ignore
-						!changeEvent.changed[0]?.input?.uri ||
-						// @ts-ignore
-						!changeEvent.changed[0]?.input?.uri.path) return;
-					// @ts-ignore
-					const filePath = changeEvent.changed[0]?.input?.uri.path;
+					const changedPath = changeEvent.changed[0] as ITab;
+					if (!changedPath?.input?.uri?.path) return;
+					const filePath = changedPath.input.uri.path;
```
Comment (ZahidSaeed01) [inline src/extension.ts:103]: 👌

File: src/settings.ts [hunk 5]
```diff
@@ -1,24 +1,32 @@
...
-    CODESYNC_HOST: useStaging ? 'https://api-staging.codesync.com': 'http://localhost:8000',
-    WEB_APP_URL: useStaging ? "https://staging.codesync.com": "http://localhost:3000"
+    ROOT_REPO: useStaging ? stagingConfig.ROOT_REPO: devConfig.ROOT_REPO,
+    API_HOST: useStaging ? stagingConfig.API_HOST: devConfig.API_HOST,
+    WEBAPP_HOST: useStaging ? stagingConfig.WEBAPP_HOST: devConfig.WEBAPP_HOST,
+    WEBSOCKET_HOST: useStaging ? stagingConfig.SOCKET_HOST: devConfig.SOCKET_HOST,
+    CW_LOGS_GROUP: useStaging ? stagingConfig.CW_LOGS_GROUP: devConfig.CW_LOGS_GROUP,
+    AWS_REGION: useStaging ? stagingConfig.AWS_REGION: devConfig.AWS_REGION,
 };
+export const API_HOST = DEBUG ? DevConfig.API_HOST: prodConfig.API_HOST;
```
Comment (ZahidSaeed01) [inline src/settings.ts:18]: 👍

File: src/settings.ts [hunk 6]
```diff
@@ -72,6 +80,7 @@ export const generateSettings = () => {
         LOCKS_REPO: path.join(rootRepo, ".locks"),
         S3_UPLOADER: path.join(rootRepo, ".s3_uploader"),
         CONFIG_PATH: path.join(rootRepo, "config.yml"),
+        ON_PREM_CONFIG: path.join(rootRepo, "on_prem_config.json"),
```
Comment (ZahidSaeed01) [inline src/settings.ts:83]: 👌

**Review by ZahidSaeed01 — APPROVED** [review]
Amazing work @Basit-qc  🙌

**Pull request merged**